keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -656,73 +661,287 @@ ways.  One is that we are dispatching requests to be served rather
+```
+
+(R and max commute because both are monotonically non-decreasing).
+
+Note that `mu_fair(t)` is exactly the same as `dR/dt` in the original
+story.  So we can reason as follows.
+
+```
+Integral[tau=B(i,j) to tau=E(i,j)] mu(i,tau) dtau = len(i,j)
+
+Integral[tau=start to tau=E(i,j)] (dR/dt)(tau) dtau -
+Integral[tau=start to tau=B(i,j)] (dR/dt)(tau) dtau
+= len(i,j)
+
+R(E(i,j)) - R(B(i,j)) = len(i,j)
+
+F(i,j) - S(i,j) = len(i,j)
+
+F(i,j) = S(i,j) + len(i,j)
+```
+
+It is not necessary to track the B and E values for every unsent
+packet in the virtual world.  It suffices to track, for each queue
+`i`, the following things in the virtual world:
+- the contents of the queue
+- B of the packet currently being sent
+- bits of that packet sent so far: `Integral[from tau=B to tau=now] mu(i,tau) dtau`
+
+At any point in time the E of the packet being sent can be calculated,
+under the assumption that `mu(i,t)` will henceforth be constant, by
+adding to the current time the quotient of (remaining bits to send) /
+`mu(i,t)`.  The E of the packet after that can be calculated by
+further adding the quotient (size of packet) / `mu(i,t)`.  This E is
+the value used to pick the next packet to send in the real world.
+
+If we are satisfied with an O(num queues) cost to react to advancing
+the clock or picking a request to dispatch then a direct
+implementation of the above works.
+
+It is possible to reduce those costs to O(log(num queues)) by
+leveraging the above correspondence to work with R values and keep the
+queues in a data structure sorted by F of the next packet to transmit
+in the virtual world.  There will be two classes of queues: those that
+do not and those that do have a packet waiting to start transmission
+in the virtual world.  It is the latter class that is kept in a data
+structure sorted by the F of that packet.  Such a packet's F does not
+change as the system evolves over time, so an incremental step
+requires only manipulating the queue and packet directly involved.
+
+##### From one to many
+
+The first change is to suppose that transmission is being done on
+multiple parallel links.  Call the number of them `C`.  This allows
+`mu_fair` to be higher.  Its constraint changes to
+
+```
+min(C*mu_single, Sum[over i] rho_i) = Sum[over i] min(rho_i, mu_fair)
+```
+
+Because we now have more possible values for `mu(i,t)` than 0 and
+`mu_fair(t)`, it is more computationally complex to adjust the
+`mu(i,t)` values when a packet arrives or completes virtual service.
+That complexity is:
+- O(n log n), where n is the number of queues,
+  in a straightforward implementation;
+- O(log n) if the queues are kept in a data structure sorted by `rho(i,t)`.
+
+We can keep the same virtual transmission scheduling scheme as in the
+single-link world --- that is, each queue sends one packet at a time
+in the virtual world.  This looks a little unreal but that is
+immaterial and keeps the logic simple; we can use the same B and E
```

Comment:
I'm not convinced about this step. Do you intend for a single queue to serve more than one request at a time? It sounds like no, but then there's two scenarios that seem problematic:
a) There are C links, only one queue has work to do, but that queue has C (or more) pending packets
b) Fewer than C queues are busy; one or more queues have many waiting packets, but the one(s) in front of the queue will take 5 minutes to serve, and the subsequent ones will take 1 second to serve.

It seems like either scenario would end up serving many fewer requests than the system can actually handle?

However, if you permit a queue to serve more than one request at a time, it seems like the math needs some changes since packet service begin/end times in a single queue will no longer be non-overlapping.

I suspect addressing this point will accidentally solve the scenario I've been worried about for long-running requests. I'd love to take a stab at fixing the math up to handle serving more than one packet at a time but I suspect I won't have a time block soon :(

Reply:
Also I think a TODO here is fine, because what you've got here is an improvement already, it might be good to merge it and then further improve it later.

Reply:
I added text that I hope addresses the concerns expressed here.
I do not think that this is going to handle long-running requests the way we want.  I think we want noticeably different handling for them.